Don't reschedule syncs on startup

dpsutton opened the pull request:
Fixes #22650 

Previously when we started up, for each database we compare its current
schedule to its schedule in Quartz, and update the triggers if they were
different. But we handle this change on db inserts and db updates so
this is completely unnecessary. We did this back when quartz was in
memory only and it remained when we persisted.

We can remove this complexity. On an app-db with 400 postgres databases,
this takes 46 seconds in the old world and appears as just another log
statement now.

Reports from the wild are even more distressing: the quartz tables got
full of dead tuples causing queries to the triggers to take a long
time. They also had 500+ databases. We query all triggers and linear
search for each database and this process exploded up to as much as 8
hours. Now we are not doing all of this extra work and startup should be
quick again.

dpsutton committed:
Don't reschedule syncs on startup

Previously when we started up, for each database we compare its current
schedule to its schedule in Quartz, and update the triggers if they were
different. But we handle this change on db inserts and db updates so
this is completely unnecessary. We did this back when quartz was in
memory only and it remained when we persisted.

We can remove this complexity. On an app-db with 400 postgres databases,
this takes 46 seconds in the old world and appears as just another log
statement now.

Reports from the wild are even more distressing: the quartz tables got
full of dead tuples causing queries to the triggers to take a long
time. They also had 500+ databases. We query all triggers and linear
search for each database and this process exploded up to as much as 8
hours. Now we are not doing all of this extra work and startup should be
quick again.

dpsutton committed:
clj-kondo ignore unused private var

dpsutton committed:
appease ns linter

dpsutton committed:
remove clojure.core/binding and unused bindings

camsaul commented:
I don't think this really answers the question of how the DB got full of dead tuples in the first place... theoretically the code is not supposed to be deleting/inserting rows if the schedule hasn't changed. Was that code broken and did it do delete/recreate on every launch?

camsaul commented on src/metabase/task/sync_databases.clj:
```diff
@@ -206,6 +206,8 @@
       (cron/with-misfire-handling-instruction-do-nothing)))))
 
+;; called [[from metabase.models.database/schedule-tasks!]] from the post-insert and the pre-update
+#_ {:clj-kondo/ignore [:unused-private-var]}
 (s/defn ^:private check-and-schedule-tasks-for-db!
```
If this is used in another namespace we should probably just make it public. Otherwise someone might think "hey it's private, I can rename it or delete it or whatever" and it's going to cause confusion when that breaks something

dpsutton replied:
```
    ;; this is done this way to avoid circular dependencies
    (classloader/require 'metabase.task.sync-databases)
    ((resolve 'metabase.task.sync-databases/check-and-schedule-tasks-for-db!) database)
```
it's used but there (or perhaps were?) circularity issues.

camsaul replied:
Yeah. I think we should make it public because it's used elsewhere even if we aren't using it directly without dumb `resolve` hacks

camsaul commented on src/metabase/task/sync_databases.clj:
```diff
@@ -254,37 +256,30 @@
   (task/add-job! sync-analyze-job)
   (task/add-job! field-values-job))
 
-(defn randomized-schedules
-  "Updated default schedules for the sync task when given the original default schedules.
-  The defaults used to be all at the same time leading to poor resource management. If the user has not indicated they
-  want to control the scheduling, returns appropriate randomized schedules for the sync tasks."
-  [{:keys [details cache_field_values_schedule metadata_sync_schedule] :as _database}]
-  (when-not (:let-user-control-scheduling details)
-    (let [random-defaults (sync.schedules/schedule-map->cron-strings (sync.schedules/default-schedule))
-          old-default-keys (cond-> []
-                             (contains? sync.schedules/default-cache-field-values-schedule-cron-strings cache_field_values_schedule)
-                             (conj :cache_field_values_schedule)
-
-                             (contains? sync.schedules/default-metadata-sync-schedule-cron-strings metadata_sync_schedule)
-                             (conj :metadata_sync_schedule))]
-      (not-empty (select-keys random-defaults old-default-keys)))))
```
Just to check, we're still randomizing the schedules when we create a new Database, right? IIRC you added that in #14343 so we don't blast people with a ton of simultaneous syncs or something like that

dpsutton replied:
Should be. We are still randomizing them on startup for any that are default in this pr:
```clojure
(defmethod task/init! ::SyncDatabases
  [_]
  (job-init)
  ;; prevent "stampedes" on databases which all used the same sync schedules. Randomize them
  (let [dbs-with-default-schedules (->> (db/select Database)
                                        (filter
                                         (every-pred
                                          metabase-controls-schedule? old-default-schedule?)))]
    (when (seq dbs-with-default-schedules)
      (log/info (trs "Updating databases without custom sync scheduling to random schedules.")))
    (doseq [db dbs-with-default-schedules]
      (db/update! Database (u/the-id db)
        (sync.schedules/schedule-map->cron-strings (sync.schedules/default-schedule))))))
```

So we still have this code present which runs on startup to essentially migrate preexisting dbs to use a randomized schedule. I'm gonna track down the code that ensures dbs start with a randomized schedule as well.

dpsutton replied:
and yes new databases are given random schedules on line 598 of api/database: https://github.com/metabase/metabase/blob/master/src/metabase/api/database.clj#L598

camsaul replied:
Nvm I see the code wherer that's happening below

camsaul commented on src/metabase/task/sync_databases.clj:
```diff
@@ -254,37 +256,30 @@
+                                          metabase-controls-schedule? old-default-schedule?)))]
+    (when (seq dbs-with-default-schedules)
+      (log/info (trs "Updating databases without custom sync scheduling to random schedules.")))
+    (doseq [db dbs-with-default-schedules]
+      (db/update! Database (u/the-id db)
+        (sync.schedules/schedule-map->cron-strings (sync.schedules/default-schedule))))))
```
Oh ok is this where the randomization happens? And it should no-op after the first update, right?

camsaul replied:
Can you add a test to make sure this is no-opping? Just check that the `updated_at` timestamp changes the first time but not the second or something like that

dpsutton replied:
yes exactly.

camsaul commented on src/metabase/task/sync_databases.clj:
```diff
@@ -254,37 +256,30 @@
-      (catch Throwable e
-        (log/error e (trs "Failed to schedule tasks for Database {0}" (:id database)))))))
+  ;; prevent "stampedes" on databases which all used the same sync schedules. Randomize them
+  (let [dbs-with-default-schedules (->> (db/select Database)
+                                        (filter
+                                         (every-pred
+                                          metabase-controls-schedule? old-default-schedule?)))]
```
It seems super inefficient to fetch the entire row for _every_ Database on every launch like this. You could extremely easily only select the fields you need, e.g.

```clj
(db/select [Database :id :details ...])
```

You can also extremely easily filter do `old-default-schedule?` directly in SQL with something like

```clj
(db/select [Database :id :details ...]
           {:where [:or
                    [:in :cache_field_values_schedule default-cache-field-values-schedule-cron-strings]
                    [:in :metadata_sync_schedule default-metadata-sync-schedule-cron-strings]]})
```

Unfortunately `old-default-schedule?` still has to be done in Clojure-land but once #22354 gets merged (and it seems like [@bshepherdson](https://github.com/bshepherdson) is really close on getting it ready) we can do the entire thing in Database-land. At which point in the normal happy-path `db/select` would return no rows.

The only reason I'm mentioning this is there can be lots of Databases in big instances and fetching 1000 entire rows over the wire and doing all the Toucan magic can be a little expensive. Doing things in the Database is way faster so if we can do it there we should

camsaul replied:
I would also suggest using `db/select-reducible` and reducing/transducing the results so we're handling each row as it comes and then tossing it instead of potentially fetching 1000 rows at once and then iterating over all of them. That should make this faster and reduce the memory impact

dpsutton replied:
I'm 100% on board with this. I'm not following why `old-default-schedule?` cannot be done in database land given that you gave an example above it. I think you're talking json access for `:let-user-control-scheduling` check for `metabase-controls-schedule?`? If so I'm on board with these changes.

camsaul replied:
Oh yeah sorry that's what I meant. `let-user-control-scheduling` can't be done in Clojure-land until we get the H2 upgrade in so we have native JSON functions. 

Well actually if you wanted to be hacky I suppose it could be done in Database-land for Postgres and MySQL and in Clojure-land for H2 until #22354 gets merged in. Maybe a bit too much work. It would be nice to add a comment tho or open a GH issue so we remember to come change it over later

dpsutton committed:
Rename `default-schedules` -> `default-randomized-schedule`

makes it a bit more clear what's going on

camsaul reviewed (approved):
Tentatively approving but I really think you should try to optimize that code that selects and randomizes DB queries on launch a bit. It seems like an easy way to shave off a lot of init time for larger instances

dpsutton committed:
Put randomizer in transducer over reducible query; tests

basically a `run!` over a reducible query that also does a count for
logging. Tests now check the three options:
- default schedules, updated
- non-default schedules, not updated timestamp verified that it is
identical.
- default schedules but user controls schedule: not updated, timestamp
verified that it is identical.

dpsutton commented:
thanks for the review @camsaul . I've switched to a reducible query so we're updating on a pull model from the query results rather than getting them all in memory. Thanks for the review.

dpsutton committed:
Prevent randomized schedules from returning a default value

it actually doesn't matter in the application but the tests are
asserting that the newly randomized schedule is not one of the defaults
it was changed from and that can fail sometimes. Doesn't matter
logically but no biggie to simplify the tests.

Otherwise they fail in (1/60)+(1/24) test runs which is actually
decently frequently with 26 test suites per commit
